### PR TITLE
Proper  initialization of DLInfo

### PR DESCRIPTION
**Pull request by khagankhan**
Fixed https://github.com/WasmEdge/WasmEdge/issues/3014

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

File: lib/plugin/plugin.cpp [hunk 1]
```diff
@@ -273,6 +273,10 @@ std::vector<std::filesystem::path> Plugin::getDefaultPluginPaths() noexcept {
                        std::filesystem::u8path("plugin"sv));
     }
   }
+  else {
+    fprintf(stderr, "Error: dladdr failed to find the shared object or symbol.\n");
```
[the remark this answers is not in the document]
Comment (khagankhan) [reply lib/plugin/plugin.cpp:277]: Gotcha!

File: lib/plugin/plugin.cpp [hunk 2]
```diff
@@ -250,13 +250,13 @@ std::vector<std::filesystem::path> Plugin::getDefaultPluginPaths() noexcept {
   Dl_info DLInfo;
   int Status =
       dladdr(reinterpret_cast<void *>(Plugin::getDefaultPluginPaths), &DLInfo);
-  if (Status != 0) {
-    auto LibPath = std::filesystem::u8path(DLInfo.dli_fname)
-                       .parent_path()
-                       .lexically_normal();
+  if (Status != 0 && DLInfo.dli_fname != nullptr) {
+    auto initialPath = std::filesystem::u8path(DLInfo.dli_fname);
```
[the remark this answers is not in the document]
Comment (khagankhan) [reply lib/plugin/plugin.cpp:254]: Okay!

File: lib/plugin/plugin.cpp [hunk 3]
```diff
@@ -273,6 +273,10 @@ std::vector<std::filesystem::path> Plugin::getDefaultPluginPaths() noexcept {
                        std::filesystem::u8path("plugin"sv));
     }
   }
+  else {
+    fprintf(stderr, "Error: dladdr failed to find the shared object or symbol.\n");
+    exit(EXIT_FAILURE);
```
[the remark this answers is not in the document]
Comment (khagankhan) [reply lib/plugin/plugin.cpp:278]: Got it!

File: lib/plugin/plugin.cpp [hunk 4]
```diff
@@ -250,13 +250,13 @@ std::vector<std::filesystem::path> Plugin::getDefaultPluginPaths() noexcept {
...
   int Status =
       dladdr(reinterpret_cast<void *>(Plugin::getDefaultPluginPaths), &DLInfo);
-  if (Status != 0) {
-    auto LibPath = std::filesystem::u8path(DLInfo.dli_fname)
-                       .parent_path()
-                       .lexically_normal();
+  if (Status != 0 && DLInfo.dli_fname != nullptr) {
+    auto initialPath = std::filesystem::u8path(DLInfo.dli_fname);
+    auto parentPath = initialPath.parent_path();
+    auto LibPath = parentPath.lexically_normal();
```
[the remark this answers is not in the document]
Comment (khagankhan) [reply lib/plugin/plugin.cpp:256]: Oh! Probably forgot after debugging!

**Comment by khagankhan** [comment]
I will address the comments asap and will update it! Thanks!

**Comment by dannypsnl** [comment]
@khagankhan Do you have any updates?

**Comment by khagankhan** [comment]
Hey! Sorry for that, it is end of the semester so I have been busy with the that of late. I will have done the necessary updates by the end of this week!

**Review by khagankhan — COMMENTED** [review]
Addressed the change requests. MSAN report does not appear for the `DLInfo.dli_fname` because it is now checked before used. The error message is used with `spdlog::error`. I may look at the usage of `parental_path()` and `lexically_normal()` down the road because I ran into something that may be interesting.

File: lib/plugin/plugin.cpp [hunk 5]
```diff
@@ -272,6 +272,8 @@ std::vector<std::filesystem::path> Plugin::getDefaultPluginPaths() noexcept {
       Result.push_back(LibPath / std::filesystem::u8path(".."sv) /
                        std::filesystem::u8path("plugin"sv));
     }
+  } else {
+    spdlog::error("Error: dladdr failed to find the shared object or symbol.");
```
Comment (dannypsnl) [inline lib/plugin/plugin.cpp:276]: You will need to abort the process, @hydai do you know how we usually do this?

**Review by dannypsnl — CHANGES_REQUESTED** [review]

File: lib/plugin/plugin.cpp (hunk 5 above)
Comment (dannypsnl) [reply lib/plugin/plugin.cpp:276]: Except that I think the PR is ok

**Comment by dannypsnl** [comment]
You will need to sign your commit, which means each commit message will have a email line, and you can use rebase master to squash the commits.

**Comment by khagankhan** [comment]
Hey, if it needs to abort after the error message. I can commit and sign it subject to the approval!

Best, 

-Khan

**Comment by dannypsnl** [comment]
Sure, please do so.

File: lib/plugin/plugin.cpp [hunk 6]
```diff
@@ -250,7 +250,7 @@ std::vector<std::filesystem::path> Plugin::getDefaultPluginPaths() noexcept {
   Dl_info DLInfo;
   int Status =
       dladdr(reinterpret_cast<void *>(Plugin::getDefaultPluginPaths), &DLInfo);
-  if (Status != 0) {
+  if (Status != 0 && DLInfo.dli_fname != nullptr) {
```
Comment (hydai) [inline lib/plugin/plugin.cpp:253]: We should have this condition inside the `Status != 0`.
You can find that the `else` only handles the `DLInfo.dli_fname == nullptr` with the given error message. However, one possible failure is that the `dladdr` fails and returns 0.

**Review by hydai — CHANGES_REQUESTED** [review]

**Review by dannypsnl — APPROVED** [review]

**Comment by dannypsnl** [comment]
@khagankhan Please sign-off your commits, so @hydai can merge it.

**Comment by khagankhan** [comment]
Hey! Thank you for the addressing the last comment! I will make it ready during the day. Thanks, guys! Appreciate it!

**Comment by khagankhan** [comment]
rebased the branch to sign off the previously unsigned commits.

**Comment by khagankhan** [comment]
I absolutely have no idea how these commits appeared. If needed to resolve the conflicts, I can close and make another PR with just one commit to resolve the bug.

**Comment by khagankhan** [comment]
A new PR is made. This PR is hard to resolve and does not reflect the exact problem, solution, and the purpose.

**Comment by hydai** [comment]
If you are rebasing to the master, these commits should not happen because all of them are in the commits history of the master branch if you have no idea how to fix it. Please check out a new branch from the master branch and cherry-pick your commits into the new branch.

**Comment by khagankhan** [comment]
Absolutely! It is about time to make a  new PR after carefully addressing each comment in a single commit. Sorry for inconvenience and thank you for the help!

**Pull request closed without merging**